Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
                                  1 -1 -1  1  1 -1 -1  1;
                                  1  1 -1 -1  1  1 -1 -1;
                                  1 -1  1 -1  1 -1  1 -1];
-		end    
-    end
+		end
+	end
     
-    if strcmp(xQ.Vendor,'Philips')
-        xQ.TimeEncodedMatrix = xQ.TimeEncodedMatrix * -1;
+    switch xQ.Vendor
+        case 'Siemens'
+            % This is where this code was initially based on, so for
+            % Siemens we don't do anything specific
+        case 'Philips'
+            xQ.TimeEncodedMatrix = xQ.TimeEncodedMatrix * -1;
+            % Philips uses an inverse matrix compared to Siemens
+        case 'GE'
+            warning('Time encoded ASL data detected for GE, but GE usually does the decoding in K-space on the scanner');
+        otherwise
+            warning(['Time encoded ASL data not yet tested for vendor ' xQ.vendor]);
     end
     
 end
 
-% Load time-series nifti
+
+%% 2. Load time-series nifti
 imEncoded = xASL_io_Nifti2Im(imPath); 
 
-% Calculate specific data sizes
-nDecodedTI = xQ.TimeEncodedMatrixSize-1;                   % Number of TIs is always MatrixSize -1
-nDecodedVolumes = xQ.NumberEchoTimes * nDecodedTI;
+% Prepare the general sequence parameters
+nEncodedVolumes = size(imEncoded, 4);     % Size of the raw data
+
+% Check that input quantification parameters are vectors with an equal length to the number of volumes
+if length(xQ.LabelingDuration)>1 && length(xQ.LabelingDuration) ~= nEncodedVolumes
+	error('Length of LabelingDuration vector does not match the number of volumes.');
+end
+
+if length(xQ.Initial_PLD)>1 && length(xQ.Initial_PLD) ~= nEncodedVolumes
+	error('Length of Initial_PLD vector does not match the number of volumes.');
+end
+
+% Prepare the parameters also for interleaved reordering
+% We thus calculate the size of each Hadamard block as the number of Hadamard phases and TEs
+nVolumesPerRepetition = xQ.TimeEncodedMatrixSize * xQ.NumberEchoTimes;
+nRepetitions = nEncodedVolumes / nVolumesPerRepetition; % Calculating no. of acquisition repeats accounting for multipleTE
+
+if length(xQ.EchoTime)>1 && length(xQ.EchoTime) ~= nEncodedVolumes
+	error('Length of EchoTime vector does not match the number of volumes.');
+end
+
+% Currently, we only implement the option that TE times are together
+% In case a different ordering is used, an error is reported
+if isfield(xQ, 'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
+	% The first two EchoTimes are equal
+	if length(uniquetol(xQ.EchoTime(1:2), 0.001)) == 1
+		error('Processing of multi-TE sequence is only implemented for the case of TEs being together in ASL4D.')
```

Comment:
Can you improve this error message, first you say that this function only works for the first two echotimes being equal, then you say "being together in ASL4D". I don't understand what this means.

Also, we just concluded that we use the innerDim and not the EchoTimes. So if there is any incompatibility of EchoTime with Hadamard decoding, this should give a warning but not crash the decoding

Reply:
Essentially - we only implement the option PLD1TE1, PLD1TE2, PLD1TE3, PLD2TE1. The option PLD1TE1, PLD2, TE1, PLD3TE1, PLD1TE2 is not implemented and issues an error. So the code is correct but the comments are indeed confusing as they look contradictory. They are not really contradictory, just badly written. I'll improve that.

We can go through innerDims indeed. But if we find (by PLD count) that there are no innerDims. But we know that there are several TEs but not as innerDims, then we report an errors. Because we need to deal with TE position and this option when TEs are not inner is not implemented and tested.